sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatch.java

```diff
@@ -0,0 +1,322 @@
+        long offset = keyOffsets[rowId];
+        klen = Platform.getInt(currentAndOnlyBase, offset - 4);
+        keyRow.pointTo(currentAndOnlyBase, offset, klen);
+      }
+      // set keyRowId so we can check if desired row is cached
+      keyRowId = rowId;
+    }
+    return keyRow;
+  }
+
+  /**
+   * Returns the value row in this batch at `rowId`.
+   * It can be a faster path if `keyRowId` is equal to `rowId`, which means the preceding
+   * key row has just been accessed. This is always the case so far.
+   * Returned value row is reused across calls.
+   */
+  public UnsafeRow getValueRow(int rowId) {
+    return getValueFromKey(rowId);
+  }
+
+  /**
+   * Returns the value row in this batch at `rowId`.
+   * It can be a faster path if `keyRowId` is equal to `rowId`, which means the preceding
+   * key row has just been accessed. This is always the case so far.
+   * Returned value row is reused across calls.
+   */
+  private UnsafeRow getValueFromKey(int rowId) {
+    if (keyRowId != rowId) {
+      getKeyRow(rowId);
+    }
+    assert(rowId >= 0);
+    if (allFixedLength) {
+      valueRow.pointTo(currentAndOnlyBase,
+              keyRow.getBaseOffset() + klen,
+              vlen + 4);
+    } else {
+      long offset = keyOffsets[rowId];
+      vlen = Platform.getInt(currentAndOnlyBase, offset - 8) - klen - 4;
+      valueRow.pointTo(currentAndOnlyBase,
+              offset + klen,
+              vlen + 4);
+    }
+    return valueRow;
+  }
+
+  public long spill(long size, MemoryConsumer trigger) throws IOException {
+    throw new OutOfMemoryError("row batch should never spill");
+  }
+
+  /**
+   * Returns an iterator to go through all rows
+   */
+  public org.apache.spark.unsafe.KVIterator<UnsafeRow, UnsafeRow> rowIterator() {
+    return new org.apache.spark.unsafe.KVIterator<UnsafeRow, UnsafeRow>() {
+      private final UnsafeRow key = new UnsafeRow(keySchema.length());
+      private final UnsafeRow value = new UnsafeRow(valueSchema.length());
+
+      private long offsetInPage = 0;
+      private int recordsInPage = 0;
+
+      private int currentklen;
+      private int currentvlen;
+      private int totalLength;
+
+      private boolean inited = false;
+
+      private void init() {
+        if (currentAndOnlyPage != null) {
+          offsetInPage = currentAndOnlyPage.getBaseOffset();
+          recordsInPage = Platform.getInt(currentAndOnlyBase, offsetInPage);
+          offsetInPage += 4;
+        }
+        inited = true;
+      }
+
+      @Override
+      public boolean next() {
+        if (!inited) init();
+        //searching for the next non empty page is records is now zero
+        if (recordsInPage == 0) {
+          freeCurrentPage();
+          return false;
+        }
+
+        if (allFixedLength) {
+          totalLength = klen + vlen + 4;
+          currentklen = klen;
+          currentvlen = vlen;
+        } else {
+          totalLength = Platform.getInt(currentAndOnlyBase, offsetInPage);
```

Comment:
I'm not entirely sure why the bytes to bytes map used `klen + vlen + 4` in the first place, but it seemed intentional (@rxin do you know?). It could have something to do with the spilling logic in which case it should be safe to remove here. But if we don't find the reason, perhaps we can just add 4 to be consistent with the previous implementation for now and revisit this in future.